Source/Core/InputCommon/ControllerInterface/evdev/evdev.cpp

```diff
@@ -26,22 +38,112 @@ static std::string GetName(const std::string& devnode)
+
+  struct udev* udev = udev_new();
+  _assert_msg_(PAD, udev != nullptr, "Couldn't initilize libudev.");
+
+  // Set up monitoring
+  struct udev_monitor* monitor = udev_monitor_new_from_netlink(udev, "udev");
+  udev_monitor_filter_add_match_subsystem_devtype(monitor, "input", nullptr);
+  udev_monitor_enable_receiving(monitor);
+  int monitor_fd = udev_monitor_get_fd(monitor);
+
+  while (s_hotplug_thread_running.IsSet())
+  {
+    fd_set fds;
 
-  int num_controllers = 0;
+    FD_ZERO(&fds);
+    FD_SET(monitor_fd, &fds);
+    FD_SET(s_pipe_fds[0], &fds);
 
-  // We use Udev to find any devices. In the future this will allow for hotplugging.
-  // But for now it is essentially iterating over /dev/input/event0 to event31. However if the
+    int ret = select(monitor_fd + 1, &fds, nullptr, nullptr, nullptr);
+    if (ret < 1 || !FD_ISSET(monitor_fd, &fds))
+      continue;
+
+    udev_device* dev = udev_monitor_receive_device(monitor);
+
+    const char* action = udev_device_get_action(dev);
+    const char* devnode = udev_device_get_devnode(dev);
+
+    if (!devnode)
+      continue;
+
+    // Only react to "device add" events for evdev devices that we can access.
+    // If the device name is empty, it is probably not an evdev device, so we skip it
+    // Note: evdev can actually return an empty string, but in such cases,
+    // we have nothing to match against when trying to detect which device
+    // was unplugged then replugged.
+    std::string name = GetName(devnode);
+    if (strcmp(action, "add") == 0 && access(devnode, W_OK) == 0 && !name.empty())
+    {
+      NOTICE_LOG(SERIALINTERFACE, "Found new device: %s (%s)", devnode, name.c_str());
+      // Try to find a device that was unplugged then replugged
+      // If found, update the fd on the libevdev device, so we can get input from the device again.
+      std::shared_ptr<ciface::Core::Device> d = g_controller_interface.FindDevice(name, "evdev");
+      if (d != nullptr)
+      {
+        auto device = std::static_pointer_cast<evdevDevice>(d);
+
+        if (device->IsStillValid())
+          continue;
+
+        int new_fd = open(devnode, O_RDWR | O_NONBLOCK);
+        device->ChangeFd(new_fd);
+        NOTICE_LOG(SERIALINTERFACE, "Detected replugged device (%s): fd changed to %d",
+                   name.c_str(), new_fd);
+      }
+      else
+      {
+        auto input = std::make_shared<evdevDevice>(devnode, s_name_counts[name]++);
+
+        if (input->IsInteresting())
+        {
+          g_controller_interface.AddDevice(std::move(input));
+          NOTICE_LOG(SERIALINTERFACE, "Added new device: %s", name.c_str());
+          g_controller_interface.InvokeDeviceCallbacks();
+        }
+      }
+    }
+    udev_device_unref(dev);
+  }
+  NOTICE_LOG(SERIALINTERFACE, "evdev hotplug thread stopped");
+}
+
+void StartHotplugThread()
+{
+  if (s_hotplug_thread_running.IsSet())
+    return;
+
+  pipe(s_pipe_fds.data());
```

Comment:
oh, I thought the standard way was to use pipes, as suggested by phire and a quick search. Didn't know about eventfd, which is supposedly more lightweight; will switch to it.

I have looked at using Common::Event, but it looks like select() would be easier when we're really just waiting on file descriptors.